### ISSUE TITLE
Optimum parallelisation settings helper function

### ISSUE THREAD

**Issue by IanHeywood** [ratt-ru/CubiCal#131]
Something that might come naturally to some but not to me is how to optimally set up the `--dist-ncpu` and `--dist-min-chunks` option, depending on the specs of the computer and the MS.

I'd naively assume min-chunks would be approximately the total RAM divided by the memory footprint of the (f * N_bl * N_corr * freq-chunk * time-chunk) array, rounded down to the value that's divisible by NCPU-1 (where f is some factor that accounts for my total lack of understanding about internal temporary arrays, multiple directions etc.).

Usually I err on the side of caution and let the program just run, watching htop for a bit to have a stab at what improvements I can make next time.

But, how about an 'advise' switch, where cubical looks at the parset, looks at the MS, and has a stab at guessing NCPU and min-chunks based on the available RAM and CPUs?

Straight to the bottom of the priority pile, if it's even feasible, but I think it would be a friendly, time-saving addition.

Cheers.

**Comment by JSKenyon** [comment]
This is a good suggestion. Getting maximum throughput is a bit of a dark art at the moment. I am not sure it is really possible to infer the best parameters from just the hardware, but I think it might be possible to put up a warning/suggestion box containing useful info. 

Interestingly, I have just been doing some benchmarking which shows that, for DI calibration reading MODEL_DATA, you rapidly reach a point where throwing more resources at the problem is unhelpful as you are completely dominated by reading from the MS. I have yet to test this on an SSD to see if we are still I/O bound in that case. 

My basic sanity check is that your chunk sizes should divide as evenly as possible into your data, and --dist-min-chunks must be divisible by [(--dist-ncpu) - (1)] . So for 32 chunks per tile, you want 5, 9, 17 or 33 cpus (processes). Using really large solution intervals also comes at a cost - arrays grow beyond CPU cache and performance decays. In general, I would even say that filling RAM is not the best choice as it can lead to excessive tile sizes (in the extreme case, I would still only aim at 50% of RAM per tile - CubiCal does try to load a second one while the first is being processed). The perfect case is when it takes exactly as long to populate a tile as it takes to process it.

Having written that, I should probably incorporate it into a helpful check which warns you when your parameters will lead to poorer performance.

**Comment by o-smirnov** [comment]
Yeah it's a good discussion to have, it would be good to build in some intelligence for this. It's been at the back of my mind for a while. I'm not sure we have enough experience yet for a perfect solution, but gotta start somewhere...

Related to this, we can also do internal timings to warn the user when something is sub-optimal. For example, if the main thread finds itself waiting on I/O (rather than the solvers), fewer NCPUs may be used. Etc.
